openedx/stanford/djangoapps/auth_lagunita/api.py

```diff
@@ -0,0 +1,253 @@
+        log.debug("data: %s", data)
+        result = None
+        try:
+            response = request_method(
+                url=url,
+                headers=headers,
+                data=json.dumps(data),
+            )
+        except requests.exceptions.HTTPError as exception:
+            error = "requests.exceptions: {0}".format(exception)
+        else:
+            log.info("response: %s", response)
+            log.info("response.content: %s", response.content)
+            result, error = self._parse(response)
+            if error == 'token' and not is_recursive:
+                access_token = self.refresh_access_token()
+                log.error("Refreshed access token: %s", access_token)
+                result, error = self._request(endpoint, data, method, is_recursive=True)
+        return (result, error)
+
+    def _parse(self, response):
+        """
+        Handle common parsing for HTTP responses
+        """
+        result = None
+        error = None
+        # pylint: disable=no-member
+        if not response:
+            error = 'API operation failed; no response received'
+        elif response.status_code != requests.codes.ok:
+            errors = "{code}: {message}".format(
+                code=response.status_code,
+                message=response.content,
+            )
+            error = "API operation failed; {errors}".format(
+                errors=errors,
+            )
+        else:
+            token_error_codes = [
+                '601',
+                '602',
+            ]
+            try:
+                data = response.json()
+            except ValueError as exception:
+                log.error("%s", exception)
+                data = {}
+                error = str(exception)
+            else:
+                if not data.get('success'):
+                    log.info("data: %s", data)
+                    errors = data.get('errors', [])
+                    if any(
+                            error.get('code') in token_error_codes
+                            for error in errors
+                    ):
+                        error = 'token'
+                    else:
+                        errors = '; '.join([
+                            "{code}: {message}".format(
+                                code=error.get('code', '000'),
+                                message=error.get('message', '?'),
+                            )
+                            for error in errors
+                        ])
+                        error = "API operation failed; {errors}".format(
+                            errors=errors,
+                        )
+                else:
+                    result = data['result']
+                    if len(result):
+                        result = result[0]
+                        if result.get('status') == 'skipped':
+                            result = None
+                            error = 'skipped'
```

Comment:
Do we also want to including the error `reasons`? https://developers.marketo.com/rest-api/error-codes/#record_level_errors

Reply:
I'm speaking to when this code path is reached via `subscribe` from https://github.com/Stanford-Online/edx-platform/pull/903/files#diff-127728bba4e9f71cfc07cffbe6b4ae43R109
and it receives an unrecognized `list_id`

Reply:
Steven tested and confirmed the API returns data with `success:False` and `errors` list for invalid `list_id` errors, so this is fine.